lib/Process.php

```diff
@@ -29,278 +28,131 @@ class Process {
-     * @param   string|null $cwd Working directory or use an empty string to use the working directory of the current
-     *     PHP process.
-     * @param   mixed[] $env Environment variables or use an empty array to inherit from the current PHP process.
-     * @param   mixed[] $options Options for proc_open().
+     * @param string $command Command to run.
+     * @param string $cwd Working directory of child process.
+     * @param array $env Environment variables for child process.
+     * @param array $options Options for proc_open().
+     * @param ProcessHandle $handle Handle for the created process.
      */
-    public function __construct($command, string $cwd = null, array $env = [], array $options = []) {
-        if (self::$onWindows === null) {
-            self::$onWindows = \strncasecmp(\PHP_OS, "WIN", 3) === 0;
-        }
-
-        if (\is_array($command)) {
-            $command = \implode(" ", \array_map("escapeshellarg", $command));
-        }
+    private function __construct(string $command, string $cwd, array $env, array $options, ProcessHandle $handle) {
         $this->command = $command;
-        $this->cwd = $cwd ?? "";
-
-        foreach ($env as $key => $value) {
-            if (\is_array($value)) {
-                throw new \Error("\$env cannot accept array values");
-            }
-
-            $this->env[(string) $key] = (string) $value;
-        }
-
+        $this->cwd = $cwd;
+        $this->env = $env;
         $this->options = $options;
+        $this->handle = $handle;
     }
 
     /**
      * Stops the process if it is still running.
      */
     public function __destruct() {
-        if (\getmypid() === $this->oid) {
-            $this->kill(); // Will only terminate if the process is still running.
-        }
-
-        if ($this->watcher !== null) {
-            Loop::cancel($this->watcher);
-        }
-
-        if ($this->stdin && \is_resource($resource = $this->stdin->getResource())) {
-            \fclose($resource);
-        }
-
-        if ($this->stdout && \is_resource($resource = $this->stdout->getResource())) {
-            \fclose($resource);
-        }
-
-        if ($this->stderr && \is_resource($resource = $this->stderr->getResource())) {
-            \fclose($resource);
-        }
-
-        if (\is_resource($this->process)) {
-            \proc_close($this->process);
+        if ($this->handle !== null) {
+            self::$processRunner->destroy($this->handle);
         }
     }
 
     /**
      * Resets process values.
      */
-    public function __clone() {
-        $this->process = null;
-        $this->deferred = null;
-        $this->watcher = null;
-        $this->pid = 0;
-        $this->oid = 0;
-        $this->stdin = null;
-        $this->stdout = null;
-        $this->stderr = null;
-        $this->running = false;
+    public function clone(): Promise {
+        return self::start($this->command, $this->cwd, $this->env, $this->options);
     }
```

Comment:
I'd really like to keep `start()` as normal method. Then `__clone` can just work fine, too.